test/e2e/e2e_test.go

```diff
@@ -0,0 +1,136 @@
+		podList = append(podList, pod)
+	}
+	return podList
+}
+
+// startEndToEndForLowNodeUtilization tests the lownode utilization strategy.
+func startEndToEndForLowNodeUtilization(clientset clientset.Interface) {
+	var thresholds = make(deschedulerapi.ResourceThresholds)
+	var targetThresholds = make(deschedulerapi.ResourceThresholds)
+	thresholds[v1.ResourceMemory] = 20
+	thresholds[v1.ResourcePods] = 20
+	thresholds[v1.ResourceCPU] = 70
+	targetThresholds[v1.ResourceMemory] = 20
+	targetThresholds[v1.ResourcePods] = 20
+	targetThresholds[v1.ResourceCPU] = 75
+	// Run descheduler.
+	evictionPolicyGroupVersion, err := eutils.SupportEviction(clientset)
+	if err != nil || len(evictionPolicyGroupVersion) == 0 {
+		glog.Fatalf("%v", err)
+	}
+	stopChannel := make(chan struct{})
+	nodes, err := nodeutil.ReadyNodes(clientset, "", stopChannel)
+	if err != nil {
+		glog.Fatalf("%v", err)
+	}
+	nodeUtilizationThresholds := deschedulerapi.NodeResourceUtilizationThresholds{Thresholds: thresholds, TargetThresholds: targetThresholds}
+	nodeUtilizationStrategyParams := deschedulerapi.StrategyParameters{NodeResourceUtilizationThresholds: nodeUtilizationThresholds}
+	lowNodeUtilizationStrategy := deschedulerapi.DeschedulerStrategy{Enabled: true, Params: nodeUtilizationStrategyParams}
+	ds := &options.DeschedulerServer{Client: clientset}
+	strategies.LowNodeUtilization(ds, lowNodeUtilizationStrategy, evictionPolicyGroupVersion, nodes)
+	time.Sleep(10 * time.Second)
+
+	return
+}
+
+func podScheduled(c clientset.Interface, podNamespace, podName string) wait.ConditionFunc {
+	return func() (bool, error) {
+		pod, err := c.Core().Pods(podNamespace).Get(podName, metav1.GetOptions{})
+		if err != nil {
+			// This could be a connection error so we want to retry.
+			return false, nil
+		}
+		if pod.Spec.NodeName == "" {
+			return false, nil
+		}
+		glog.Infof("pod %v got scheduled onto %v node", pod.Name, pod.Spec.NodeName)
+		return true, nil
+	}
+}
+
+// waitForPodToScheduleWithTimeout waits for a pod to get scheduled and returns
+// an error if it does not scheduled within the given timeout.
+func waitForPodToScheduleWithTimeout(cs clientset.Interface, pod *v1.Pod, timeout time.Duration) error {
+	return wait.Poll(time.Second, timeout, podScheduled(cs, pod.Namespace, pod.Name))
+}
+
+func TestE2E(t *testing.T) {
+	// If we have reached here, it means cluster would have been already setup and the kubeconfig file should
+	// be in /tmp directory.
+	clientSet, err := client.CreateClient("/tmp/admin.conf")
+	if err != nil {
+		t.Errorf("Error during client creation with %v", err)
+	}
+	if err != nil {
+		t.Errorf("Error while listing nodes with %v", err)
+	}
+	podsToBeCreated := 4
+	podList := generatePods(clientSet, podsToBeCreated)
+	// This is to ensure that all pods got scheduled.
+	for _, pod := range podList {
+		waitForPodToScheduleWithTimeout(clientSet, pod, time.Second)
+	}
+
+	startEndToEndForLowNodeUtilization(clientSet)
```

Comment:
So the pods are now scheduled without any limitation. How can it test the descheduler functionality now? My original understanding was to make one node unschedulable (so it is pod-free), schedule 4 pods to the remaining two nodes and then set the descheduling policies the way at least one of the scheduled pods must be de-scheduled. Is that no longer the intended behavior?

Reply:
So, I thought about that yesterday. The problem here is since we are not simulating cluster, rather creating pods on existing cluster, depending on the size of node, the thresholds may vary making these tests unreliable if we rely on number of pods being evicted. So, I thought of removing the checks and in the logs of tests, since I added a -v flag, we could see if atleast one pod got evicted.